### ISSUE TITLE
Stack overflow: IncrementalTin.floodFillConstrainedRegionsRecursion

### ISSUE THREAD

**Issue by jandam** [gwlucastrig/Tinfour#24]
I get Stack overflow during processing "hipped roof". When line [1] is commeted code works withou stack overflow. Tomorrow I will do more investigation. Black is polygon constraint. Orange are linear constraints

```
public static void main(String[] args) {
        double nominalPointSpacing = 0.01;
        IIncrementalTin tin = new IncrementalTin(nominalPointSpacing);
        List<IConstraint> constraintList = new ArrayList<>();

        Vertex v0 = new Vertex(-5.254566785879433, -10.120986399240792, Double.NaN);
        Vertex v1 = new Vertex(11.397855226881802, -0.2510828208178282, Double.NaN);
        Vertex v2 = new Vertex(5.243457765551284, 10.120986400172114, Double.NaN);
        Vertex v3 = new Vertex(-11.397855226648971, 0.2683861115947366, Double.NaN);

        PolygonConstraint polygon = new PolygonConstraint(v0, v1, v2, v3);
        constraintList.add(polygon);

        Vertex v4 = new Vertex(4.862010033102706, 2.8856616728007793, Double.NaN);
        Vertex v5 = new Vertex(-4.867160049267113, -2.8776928577572107, Double.NaN);

        constraintList.add(new LinearConstraint(v4, v5));
        constraintList.add(new LinearConstraint(v4, v2));
        constraintList.add(new LinearConstraint(v4, v1));
        constraintList.add(new LinearConstraint(v5, v0));
        constraintList.add(new LinearConstraint(v5, v3)); // [1] COMMENT THIS LINE

        constraintList.forEach(IPolyline::complete);

        tin.addConstraints(constraintList, false);
    }
```
![ss-stack-overflow](https://user-images.githubusercontent.com/7425758/32181480-bbe2a2b0-bd94-11e7-91e1-18245b5738da.png)

**Comment by gwlucastrig** [comment]
Martin,

Well, first off, thanks for the excellent example.   It's very helpful when we can boil down a problem to a small number of samples.

I've got to think about this some more, but here's a short term fix:

In the flood-fill routine, there are blocks of code that look like:
```
    if (!f.isConstrainedRegionMember()) {
      if (!f.isConstrained()) {
        f.setConstrainedRegionMemberFlag();
        f.setConstraintIndex(index);
      }
      floodFillConstrainedRegionsRecursion(f.getDual(), index);
    }
```

Move the line setConstrainedRegionMemberFlag before the block as shown:

```
            f.setConstrainedRegionMemberFlag();
            if (!f.isConstrained()) {
                f.setConstraintIndex(index);
            }
```

Doing so will make it so that the linear constraints are marked as members of the constrained region and will not be traversed more than once.   However, if the application code requests the constraint index for one of these edges, it will still get the correct value for the linear constraint rather than that of the flood-filled constraint.

On the other hand, I am not sure whether the linear constraints should be  treated as constrained-region members.  Their behavior isn't 100 percent consistent with other member elements.  I'll have to think about that.

Gary

P.S.  I notice that you've introduced special constructors for 2-vertex linear constraints and 4-vertex polygon constraints.  These both make sense to me and I will probably added them to the code base. Do you have any others?

**Comment by jandam** [comment]
Thank you very much for hint how to fix/workaround the issue. 
  I added new constructors just for this issue - no time to write another code.
  I will try to create some pull requests with improvements/cleanups. To simplify your work. 

Martin

30. 10. 2017 v 19:33, gwlucastrig <notifications@github.com>:

**Comment by jandam** [comment]
My suggested fix according your proposal. Skip already processed edges.
IncrementalTin:
```
  public void addConstraints(
    List<IConstraint> constraints,
    boolean restoreConformity)
  {
...
    BitSet processedSet = new BitSet(getMaximumEdgeAllocationIndex() * N_SIDES + 1);
    for (int i = 0; i < constraintList.size(); i++) {
      IConstraint c = constraintList.get(i);
      if (c.definesConstrainedRegion()) {
        ArrayList<IQuadEdge>edgesForConstraint = efcList.get(i);
        floodFillConstrainedRegion(processedSet, c, edgesForConstraint);
      }
    }
}

...

  private static void floodFillConstrainedRegion(
    final BitSet processedSet,
    IConstraint c,
    ArrayList<IQuadEdge> edgeList)
  {
    int index = c.getConstraintIndex();
    for (IQuadEdge e : edgeList) {
      if (e.isConstrainedRegionEdge()) {
        floodFillConstrainedRegionsRecursion(processedSet, e, index);
      }
    }
  }


  private static void floodFillConstrainedRegionsRecursion(final BitSet processedSet, final IQuadEdge e, final int index) {
    // There is special logic here for the case where an alternate constraint
    // occurs inside the floor-fill area. For example, a linear constraint
    // might occur inside a polygon (a road might pass through a town).
    // The logic needs to preserve the constraint index of thecontained
    // edge from the alternate constraint. In that case, the flood fill
    // passes over the embedded edge, but does not modify it.
    IQuadEdge f = e.getForward();
    int edgeIndexF = getEdgeIndex(f);
    if (!f.isConstrainedRegionMember() && !processedSet.get(edgeIndexF)) {
      processedSet.set(edgeIndexF);
      if (!f.isConstrained()) {
        f.setConstrainedRegionMemberFlag();
        f.setConstraintIndex(index);
      }
      floodFillConstrainedRegionsRecursion(processedSet, f.getDual(), index);
    }
    IQuadEdge r = e.getReverse();
    int edgeIndexR = getEdgeIndex(r);
    if (!r.isConstrainedRegionMember() && !processedSet.get(edgeIndexR)) {
      processedSet.set(edgeIndexR);
      if (!r.isConstrained()) {
        r.setConstrainedRegionMemberFlag();
        r.setConstraintIndex(index);
      }
      floodFillConstrainedRegionsRecursion(processedSet, r.getDual(), index);
    }
  }

  private static int getEdgeIndex(final IQuadEdge edge) {
    return (edge.getIndex() * N_SIDES) | edge.getSide();
  }
```

**Comment by jandam** [comment]
Similar fix for SemiVirtualIncrementalTin:

```
  @Override
  public void addConstraints(List<IConstraint> constraints, boolean restoreConformity) {

...
    BitSet processedSet = new BitSet(getMaximumEdgeAllocationIndex() * N_SIDES + 1);
    for (int i = 0; i < constraintList.size(); i++) {
      IConstraint c = constraintList.get(i);
      if (c.definesConstrainedRegion()) {
        floodFillConstrainedRegions(processedSet, c, icArray[i]);
      }
    }
...
 private void floodFillConstrainedRegions(final BitSet processedSet, IConstraint c, IntCollector intCollector) {
    int constraintIndex = c.getConstraintIndex();
    for (int i = 0; i < intCollector.n; i++) {
      IQuadEdge e = this.edgePool.getEdgeForIndex(intCollector.buffer[i]);
      if (e.isConstrainedRegionEdge()) {
        floodFillConstrainedRegionsRecursion(processedSet, e, constraintIndex);
      }
    }
  }

  private static void floodFillConstrainedRegionsRecursion(final BitSet processedSet, final IQuadEdge e, final int constraintIndex) {
    // There is special logic here for the case where an alternate constraint
    // occurs inside the floor-fill area. For example, a linear constraint
    // might occur inside a polygon (a road might pass through a town).
    // The logic needs to preserve the constraint index of thecontained
    // edge from the alternate constraint. In that case, the flood fill
    // passes over the embedded edge, but does not modify it.
    IQuadEdge f = e.getForward();
    int edgeIndexF = getEdgeIndex(f);
    if (!f.isConstrainedRegionMember() && !processedSet.get(edgeIndexF)) {
      processedSet.set(edgeIndexF);
      if (!f.isConstrained()) {
        f.setConstrainedRegionMemberFlag();
        f.setConstraintIndex(constraintIndex);
      }
      floodFillConstrainedRegionsRecursion(processedSet, f.getDual(), constraintIndex);
    }
    IQuadEdge r = e.getReverse();
    int edgeIndexR = getEdgeIndex(r);
    if (!r.isConstrainedRegionMember() && !processedSet.get(edgeIndexR)) {
      processedSet.set(edgeIndexR);
      if (!r.isConstrained()) {
        r.setConstrainedRegionMemberFlag();
        r.setConstraintIndex(constraintIndex);
      }
      floodFillConstrainedRegionsRecursion(processedSet, r.getDual(), constraintIndex);
    }
  }

  private static int getEdgeIndex(final IQuadEdge edge) {
    return (edge.getIndex() * N_SIDES) | edge.getSide();
  }
```

**Comment by jandam** [comment]
I introduced static getEdgeIndex(<edge>) to compute index. This is more general than getMarkBit/setMarkBit.

**Comment by jandam** [comment]
Modified algorithm is not throwing exception. It produces wrong results. (LogoCDT is OK)

**Comment by gwlucastrig** [comment]
Well, first off, thanks for the smart idea.  It's better than the approach I had in mind.

The problem that you are seeing  may be due to your edge index routine.  Your code is trying to treat the edges as having two flags, one for each side.  But in this case, you only want one flag indicating that the edge was visited during the flood-fill.   The build-in edge-index is defined differently for the QuadEdge and SemiVirtualEdge classes.  For QuadEdge, the same index is returned for either side of the edge. For SemiVirtualEdge, an even index is returned for the primary side (the "base reference") and an odd index is returned for the secondary side.  So you might try implementing your getEdgeIndex() routine so that, for QuadEdge, it's just a pass-through returning the index of the quad edge. For a semi-virtual edge, you would want to return the value of SemiVirtualEdge.getIndex()/2.

The original approach that I was considering was to give the edge class a "visited" flag that the algorithm could use to mark where it had processed.  But I didn't like my approach because the flags would be left in an undefined state when the flood-fill algorithm terminated.  Even if Tinfour cleaned up the flags when the algorithm was done, there would be no guarantee that other applications that used the visited flag would do so. So anyone using the visited flag would have to write defensive code to protect itself against an API that lent itself to poor coding practice.  Your approach voids that and shows good judgement in  API design.

I don't have my lab notes with me right now, so I can't look this up.... At some point I did a bunch of tests comparing the performance of BitSet versus writing to an array of integers using custom code.  The reason I though that custom code might be faster is because it could make simplifying assumptions such as the fact that you were always setting bits but never clearing them, that you knew the number of bits you needed a priori, that your index value was never out-of-bounds, etc.  As I recall, it turned out that the performance difference, if any, was smaller than I could reliably measure giving the noisy test environment that is a modern operating system.

**Comment by jandam** [comment]
There are still some artifacts (missing triangles). Even when I use following code for SemiVirtual implementation. I suppose that I have some problems with input data. I have to check it.

```
 private static int getEdgeIndex(final IQuadEdge edge) {
    return edge.getIndex() / 2;
  }
```

**Comment by gwlucastrig** [comment]
The flood-fill algorithm is something I devised myself without doing much research.  There might be something better in the published literature.  

I don't have a mathematically rigorous proof that the flood-fill reaches all triangles for all topologies, but I can't think of a counter example.  Of course, that doesn't mean much...  After all, I didn't think of your hipped-roof example either.

**Comment by gwlucastrig** [comment]
Depending on the nature of your problems, the following could be helpful or it could be extraneous information. 

Recall that if an edge is shared by two polygon constraints (which is allowed, as long as the same vertices are used for the endpoints of the edge), the flood fill algorithm does not define which constraint index will be set for the edge.   In the example image that I posted showing European countries with color-coded edges, the borders of the countries were actually rendered in gray.  The application code uses the isConstrainedRegionEdge() method to make this determination.  In practice, for shared edges the constraint-index assigned to the edge by addConstraints() method is generally an artifact of the order in which edges are constructed.  The flood-fill only guarantees the constraint index for the interior edge.

If you were color-filling triangles, and wished to determine which color to use, you should not use the constrained region edge, but should look for an interior edge. 

However, there is one case where you can get ambiguous results.  Image a case where an application specified a polygon constraint consisting of three vertices and the resulting constrained region formed a single triangle, so that there were no interior edges.  If that constrained region did not border other constrained regions, everything would be fine because the edges would be unambiguous.  But if any of the edges bordered other regions, there would be some doubt as to which constraint the enclosed area was associated with.

This last situation is a weakness in the current algorithm. One work-around for the shortcoming is to simply insert a single vertex into the TIN so that it is positioned in the interior of the triangular polygon.

**Comment by gwlucastrig** [comment]
There is a way to address the weakness I described above, though it would require a code change.  The addConstraints() method keeps an internal list of the edges that are formed as each constraint is added to the TIN.  This internal data preserves information about which side of each edge faces inward to the polygon constraint. It is used to ensure that the flood-fill operates correctly.  Currently, the internal data is discarded when the code exists addConstraints().  

As an alternate approach, we could make addConstraints return an object holding this data.  An API could be developed to expose the relevant information.

**Comment by jandam** [comment]
Attached ZIP contains constrained TriangleCollector + modified version of LogoCDT with test data and LogoPanel modified to display filled triangles.

LogoCDT displays constrained edges correctly but I'm not able to get correct triangles.
There is one line with TODO in comment. With this line there is only outline of polygon. After commenting you will get polygon with holes + linear constraints.

I suppose that I have wrong algorithm in TriangleCollector

[tinfour-collector.zip](https://github.com/gwlucastrig/Tinfour/files/1432132/tinfour-collector.zip)
![ss-collector](https://user-images.githubusercontent.com/7425758/32246921-e04be9f8-be80-11e7-888f-9c5159057752.png)

**Comment by gwlucastrig** [comment]
Thank you.  I will look at it as soon as I can.  Currently, I'm looking at
the BitSet change.

**Comment by gwlucastrig** [comment]
I did some work. Here's the result.
![constraintfill](https://user-images.githubusercontent.com/19698145/32275389-743623bc-bee1-11e7-95d6-430961ddf443.png)

The modified TriangleCollector class follows.  Due to the way edges are indexed, this class will work only for the SemiVirtualIncrementalTin.  The issue with edge indexing working differently for QuadEdge and SemiVirtualEdge has been presenting itself quite a bit lately.  I'm going to have to think about a better way to handle that (I am open to suggestions).

The main issue with the original code was in the visitTrianglesConstrained() method where it decided whether to call processEdge().    The method isConstrainedRegionEdge() indicates that an edge is the boundary of a constrained region.  Recall that, in the current state of the implementation, the constrained region edge does not carry information indicating on which side of the edge the constraint belongs. So only the interior edges are a reliable indicator.  So the new version looks for isConstrainedRegionMember() and an indicator of where to start. Recall that there is an additional complication in the current implementation: a constrained region boundary is also a constrained region member.  Thus there is the compound test in the code 

`e.isConstrainedRegionMember() && !e.isConstrainedRegionEdge()` 

The other issue that still remains is the one I described in yesterday's posts above...  There can be an ambigous case where a polygon constraint includes 3 vertices and is bordered by other polygon constraints on all three sides.  In such as case, the code below will produce a triangular hole.

I am planning on doing additional work on this issue as we learn more about the problem.  Also, as I wrote the answer above, I realized that I probably should have named the method isConstrainedRegionEdge() to be isConstrainedRegionBoundary() so that the meaning of the method would have been more clear.  Alas...

[TriangleCollector.zip](https://github.com/gwlucastrig/Tinfour/files/1434203/TriangleCollector.zip)

**Comment by jandam** [comment]
Thank you very much for partial fix.

**Comment by gwlucastrig** [comment]
I am interested in creating a permanent solution for your requirements.   If you have time and it would not require excessive effort on your part, please let me know what you are trying to accomplish.

Also, I have a question.  I am beginning to thing that, once again, I need to give my methods better names that would more clearly convey the meaning of their operations.  For example:

- isConstrainedRegionEdge() would become isOnConstrainedRegionBorder()
- isConstrainedRegionMember() would become isInConstrainedRegionInterior()

The question I have is whether the alternate names would have helped you better understand the methods when you were implementing your triangle count code? 

In terms of the impact on my user base...  As far as I know you are the only user currently working with constraints.  So if the name change does present a problem for you, there is no reason to not do it.  On the other hand, if the name change wouldn't have helped, I have less motivation to make it.

**Comment by jandam** [comment]
My idea is to use tinfour as general tin generator in all parts of my project. It means generating tins for buildings and state borders. There is no other library that is so fast and can work with high precision and stability and with such huge amount of input data. You've made excellent work. Main problem is that I don't have enough time to study and better contribute to your library. So I report issues only. I work parallel on 3 modules and my boss has another 2 modules on queue :-(.

**Comment by jandam** [comment]
Different code for floodFill in standard and semi virtual implementation:
*) semi virtual  uses different flag for forward and reverse edges.
Forward calls if (!f.isConstrainedRegionEdge... and Reverse calls if (!r.isConstrainedRegionMember()
Standard implementation calls isConstrainedRegionEdge in both cases.
*) there is another difference in code floodFill. Semi virtual uses .getIndex()|0x01 for visited index and standard uses only .getIndex().
 I suppose that code should be same after latest changes

**Comment by gwlucastrig** [comment]
The OR operation should be there, but the code probably works okay without
it. The edges only need to be visited once. So the OR ensures it gets
marked as visited no matter which side it is traversed from. The OR
improves efficiency. I will write more detail when i have access to my
computer

**Comment by jandam** [comment]
I understand that getIndex()|0x01 means that edge and its dual are marked as visited. So this performance improvement can be used in standard TIN too.

**Comment by gwlucastrig** [comment]
I've updated the indexing in both standard and semi-virtual cases to use the divide-by-two approach.

**Comment by gwlucastrig** [comment]
I've made the last of the name changes to the public methods in IQuadEdge to support the constrained region concept.  I believe that I have finally settled on method names that will be meaningful to developers writing their own applications.  I apologize for the many revisions this process required. But the names will not change again.  The method names are now

- IQuadEdge.isConstrainedRegionBorder()
- IQuadEdge.isConstrainedRegionInterior()
- IQuadEdge.isConstraiedRegionMember ()    

The Javadoc has been updated accordingly.

**Comment by gwlucastrig** [comment]
Found a bug in the recent changes.  

 In IncrementalTin (line 1869) and SemiVirtualIncrementalTin (line 1776) , the setConstrained() method incorrectly calls edge.setConstrainedRegionInteriorFlag().   This call should be removed.  I will update the code on Github later today.

**Comment by gwlucastrig** [comment]
I have added the TriangleCollector class to the Tinfour code. The triangle collector class was contributed by Martin Janda.  It provides a mechanism for traversing the TIN and collecting triangles.  One use of this new feature is illustrated in the image below.  
 
![tinfourcolorlogo](https://user-images.githubusercontent.com/19698145/32584355-dcf4296e-c4c5-11e7-888d-f7595e0f72b9.png)

To produce the image, a set of polygon constraints were generated using the Java TextLayout class.  Each of the constraints associated with a particular letter was assigned a Java Color object.  After the constraints were inserted into the TIN, the TriangleCollector was used to obtain the triangles for each constraint. The triangles were drawn with an area fill in the indicated color.   Finally, the border edges of each letter were drawn in black to outline the letters.

**Comment by gwlucastrig** [comment]
I have added the triangle collector to the LogoCDT application which is included in the test.example package shipped with Tinfour.  This latest version was used to create the colorful "Tinfour" graphic shown in my post above.

After a lot of changes, this matters cited in this Issue are now resolved.   My thanks to Martin Janda who offered a remarkable amount of support and analysis on this issue.

This issue is now closed.
